lib/charms/vault_k8s/v0/vault_client.py

```diff
@@ -24,28 +26,94 @@
 
 # Increment this PATCH version before using `charmcraft publish-lib` or reset
 # to 0 if you are raising the major API version
-LIBPATCH = 5
+LIBPATCH = 6
 
 
 logger = logging.getLogger(__name__)
 RAFT_STATE_ENDPOINT = "v1/sys/storage/raft/autopilot/state"
 
 
+@dataclass
+class Token:
+    """Class that represents token authentication for vault.
+
+    This method is the most basic and always available method to access vault.
+    """
+
+    token: str
+
+    def login(self, client: hvac.Client):
+        """Authenticate a vault client with a token."""
+        client.token = self.token
+
+
+@dataclass
+class AppRole:
+    """Class that represents approle authentication for vault.
+
+    This method is primarily used to authenticate automation programs for vault.
+    """
+
+    role_id: str
+    secret_id: str
+
+    def login(self, client: hvac.Client):
+        """Authenticate a vault client with approle details."""
+        client.auth.approle.login(role_id=self.role_id, secret_id=self.secret_id, use_token=True)
+
+
+class AuthMethod(Protocol):
```

Comment:
I don't really understand the benefit of using protocols here. I know @ghislainbourgeois recommended them but it's not clear to me what advantage this has over something like an ABC which makes the descendants explicit?

Reply:
A protocol is kinda like an interface than an ABC and Python internally uses them a lot, like `Iterable`. It's all just fancy type hints at the end of the day so I don't really mind as long as it's readable.